### ISSUE TITLE
[feature] Comparison to new GitHub Artifact Attestations?

### ISSUE THREAD

**Issue by agilgur5** [slsa-framework/slsa-github-generator#3618]
**Is your feature request related to a problem? Please describe.**
GitHub [recently released](https://github.blog/changelog/2024-05-02-artifact-attestations-public-beta/) [Artifact Attestations](https://docs.github.com/en/actions/security-guides/using-artifact-attestations-to-establish-provenance-for-builds) as a new feature. It seems to overlap with the generators here, although I'm not entirely sure.

**Describe the solution you'd like**
A comparison of the two would be good to have. I'm not sure if Artifact Attestations would entirely replace the tooling here, but could perhaps be used together with them or as a subset?

**Describe alternatives you've considered**
n/a

**Additional context**
Saw the release of this new GH feature and it seemed very simple to use. Was wondering if it would make sense to use to implement SLSA for https://github.com/argoproj/argo-workflows/issues/12099

**Comment by ramonpetgrave64** [comment]
We have not done a full evaluation yet. But for now, we know that Github's Action is meant to be at L2, while our slsa-framework's Workflows are at L3.
- https://github.com/actions/attest-build-provenance/issues/55

**Comment by laurentsimon** [comment]
I don't remember if we record the `${{ env.* }}`. I recall opening a tracking issue for it a while back but I can't find it. Env variables are another way to pass parameters to a build

**Comment by laurentsimon** [comment]
fwiw, it'd be fairly simple for a maintainer to obfuscate the code used to build their artifact by using a malicious (non-recorded) input combined with GH's script injection features :)

**Comment by agilgur5** [comment]
True. We don't use workflow parameters though, so if that's the only missing piece, it sounds like we'd be able to use Artifact Attestations for L3?

**Comment by laurentsimon** [comment]
> True. We don't use workflow parameters though, so if that's the only missing piece, it sounds like we'd be able to use Artifact Attestations for L3?

yes an no. The properties need to be project-agnostic, else there's no way for a verifier to trust builds from a builder across all their dependencies without inspecting the build workflow.

**Comment by laurentsimon** [comment]
> fwiw, it'd be fairly simple for a maintainer to obfuscate the code used to build their artifact by using a malicious (non-recorded) input combined with GH's script injection features :)

Other inputs include secrets `${{ secrets.* }}`. It's fairly common on GH for users to build and push to registry in the same trust domain (job / VM). Easy to hide malicious inputs for script injections there too. And of course there are the usual GH events that contain user-defined info (in the GH event)

It's also worth distinguishing between generators and builders:
1. [Generators](https://github.com/slsa-framework/slsa-github-generator?tab=readme-ov-file#generators) (and the GH action provenance) need to record inputs to the repository workflow (workflow inputs, vars, GH event). The generators in this repo record all of these (need to check for vars @ianlewis do you remember?), including the GH event. Secrets are still a source of inputs but we can't record it since they are secrets.
2. [Builders](https://github.com/slsa-framework/slsa-github-generator?tab=readme-ov-file#builders) need no record all the repo workflow input, secrets or vars. That's because their interface is different: we record the inputs to the reusable workflow itself, so it does not matter how it was triggered by the user: the builder builds the project and the only inputs are those from the reusable workflow. Secrets can be passed but assuming the builder itself is not subject to script injection (or uses the secrets as an input) then we have a complete set of inputs. The builder still record the GH event anyway, in case that's useful for troubleshoting

**Comment by ianlewis** [comment]
> I recall opening a tracking issue for it a while back but I can't find it. Vars variables are another way to pass parameters to a build

#1555 

> fwiw, it'd be fairly simple for a maintainer to obfuscate the code used to build their artifact by using a malicious (non-recorded) input combined with GH's script injection features :)

"features" 😄  Indeed, it's easy enough for these issues to fly under the radar that GitHub warns about it in the documentation specifically.
https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#example-of-a-script-injection-attack

For the generators, pretty sure we don't record `vars` yet unfortunately. I believe these are also not in the [event payload](https://docs.github.com/en/webhooks/webhook-events-and-payloads) so we need to add them specifically.
https://github.com/slsa-framework/slsa-github-generator/blob/2f113ffa96f45c20f09c3cd15831d24d32cffa3d/slsa/buildtype.go#L61-L65

For builders we just have normal inputs and the build itself can't read the vars unless explictly given them so can't affect the build - https://github.com/slsa-framework/slsa-github-generator/issues/1555#issuecomment-1397253544

**Comment by ianlewis** [comment]
> I would say no. The properties need to be project-agnostic, else there's no way for a verifier to trust builds from a builder across all their dependencies without inspecting the build workflow (defeats the purpose of SLSA levels). In SLSA the build level is associated with a builder and that's how we scale trust.

To be fair though, for generators, we don't do any validation on the inputs so likely likely won't make a difference in our ability to verify.

The only difference is that we can record the inputs in a trustworthy way whereas the GitHub artifact attestations currently cannot. Users of slsa-github-generator generators could implement their own policy checks over the inputs using the `--print-provenance` flag on `slsa-verifier` + cue/rego etc.

**Comment by laurentsimon** [comment]
> > I would say no. The properties need to be project-agnostic, else there's no way for a verifier to trust builds from a builder across all their dependencies without inspecting the build workflow (defeats the purpose of SLSA levels). In SLSA the build level is associated with a builder and that's how we scale trust.
> 
> To be fair though, for generators, we don't do any validation on the inputs so likely likely won't make a difference in our ability to verify.

In practice it's unlikely those inputs would be used during verification, unless the carry prod vs debug build sort of information. However they are necessary incident response.

Note that the slsa-verifier supports verifying workflow_dispatch inputs, at least for our v0.2 builders.

**Comment by ianlewis** [comment]
> In practice it's unlikely those inputs would be used during verification, unless they meaningful "high-level" info about a build, like prod vs debug. However they are necessary incident response.
> 
> Note that the slsa-verifier supports verifying workflow_dispatch inputs, at least for our v0.2 builders.

Agreed. I think that it's also likely that users would want to apply their own policy over the inputs. For example, to verify that that the prod/debug flags are set properly for production,  or to verify there are no inputs set at all etc.

**Comment by ianlewis** [comment]
FWIW, I wrote a bit about Artifact Attestations here: https://www.ianlewis.org/en/understanding-github-artifact-attestations

It's not specifically contrasting with this project but it does talk about why it's SLSA 2 and not SLSA 3. Maybe that would be helpful to folks wanting to know the differences.

**Comment by sighup1** [comment]
Hi there, I read @ianlewis blog [here](https://www.ianlewis.org/en/understanding-github-artifact-attestations)  but I'm still confused about why the Github Attest action doesn't meet SLSA build L3. Their [docs](https://docs.github.com/en/actions/how-tos/secure-your-work/use-artifact-attestations/increase-security-rating) say to move the build steps into a reusable workflow and thats enough to achieve L3. Is this incorrect? Any help appreciated 🙏
